File: paleta/pages/image_drop/image_drop_page.py
from gi.repository import Adw, GLib, Gio, Gtk, Gdk

from .dropped_image import DroppedImage
from .color_thief_panel import ColorThiefPanel
from os import path

mimes = ['text/uri-list']

@Gtk.Template(resource_path='/io/github/nate_xyz/Paleta/image_drop_page.ui')
class ImageDropPage(Adw.Bin):
    __gtype_name__ = 'ImageDropPage'

    overlay = Gtk.Template.Child(name="overlay")
    status = Gtk.Template.Child(name="status")
    thief_panel = Gtk.Template.Child(name="thief_panel")

    def __init__(self) -> None:
        super().__init__()
        self.window = None
        self.setup_drop_target()
        self.file_verified = False

    def saturate(self, window, database):
        self.window = window 
        self.thief_panel.saturate(window, database)

    def setup_drop_target(self):
        formats = Gdk.ContentFormats.new(mimes)
        drop_target = Gtk.DropTargetAsync.new(formats=formats, actions=Gdk.DragAction.COPY)
        drop_target.connect('accept', self.on_drag_accept)
        drop_target.connect('drop', self.on_drag_drop)
        
        self.overlay.add_controller(drop_target)

    def on_drag_accept(self, drop_target, drop_value):
        self.file_verified = False
        formats = drop_value.get_formats()
        if contain_mime_types(formats):
            drop_value.read_value_async(Gio.File, GLib.PRIORITY_DEFAULT, None, self.verify_file_valid)
            return True
        return False

    def verify_file_valid(self, drop, task):
        result = drop.read_value_finish(task)
        if not result:
            return
        self.file_verified = path.exists(result.get_path())

    def on_drag_drop(self, drop_target, drop_value, *args):
        if not drop_value:
            self.window.add_error_toast("Unable to read drop.")
            drop_value.finish(0)
            return False
    
        if not self.file_verified:
            self.window.add_error_toast("Unable to verify file on drop, try with the file chooser in the upper left hand corner.", 4)
            drop_value.finish(0)
            return False

        drop_value.read_value_async(Gio.File, GLib.PRIORITY_DEFAULT, None, self.load_value_async)
        return True
        
    def load_value_async(self, drop, task):
        result = drop.read_value_finish(task)
        if not result:
            self.add_error_toast("Unable to read drop.")
            drop.finish(0)
            return
        
        if self.load_image(result.get_path()):
            drop.finish(Gdk.DragAction.COPY)
        else:
            self.add_error_toast("Unable to load image.")
            drop.finish(0)

    def load_image(self, uri):
        try:
            self.thief_panel.set_image(DroppedImage(uri))
            self.status.hide()
            self.window.open_image_toast(uri)
            return True
        except Exception as e:
            print(e)
            self.window.error_image_toast(uri)
            return False

def contain_mime_types(formats):
    if formats is not None:
        return True in (formats.contain_mime_type(m) for m in mimes)
    return False

File: src/pages/image_drop/image_drop_page.rs
/* image_drop_page.rs
 *
 * SPDX-FileCopyrightText: 2023 nate-xyz
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

use adw::{prelude::*, subclass::prelude::*};
use gtk::{gdk::ContentFormats, gdk, gio, glib, glib::clone, CompositeTemplate};

use std::{cell::Cell, path::Path};
use log::{debug, error};

use crate::toasts::{add_error_toast, open_image_toast, error_image_toast};
use crate::i18n::i18n;

use super::color_thief_panel::ColorThiefPanel;
use super::dropped_image::DroppedImage;

mod imp {
    use super::*;

    #[derive(Debug, Default, CompositeTemplate)]
    #[template(resource = "/io/github/nate_xyz/Paleta/image_drop_page.ui")]
    pub struct ImageDropPagePriv {
        #[template_child(id = "overlay")]
        pub overlay: TemplateChild<gtk::Overlay>,

        #[template_child(id = "status")]
        pub status: TemplateChild<adw::StatusPage>,

        #[template_child(id = "image_bin")]
        pub image_bin: TemplateChild<adw::Bin>,

        #[template_child(id = "thief_panel")]
        pub thief_panel: TemplateChild<ColorThiefPanel>,

        #[template_child(id = "open_image_button")]
        pub open_image_button: TemplateChild<gtk::Button>,

        pub file_verified: Cell<bool>,
    }

    #[glib::object_subclass]
    impl ObjectSubclass for ImageDropPagePriv {
        const NAME: &'static str = "ImageDropPage";
        type Type = super::ImageDropPage;
        type ParentType = adw::Bin;

        fn class_init(klass: &mut Self::Class) {
            Self::bind_template(klass);
        }

        fn instance_init(obj: &glib::subclass::InitializingObject<Self>) {
            obj.init_template();
        }
    }

    impl ObjectImpl for ImageDropPagePriv {
        fn constructed(&self) {
            self.parent_constructed();
            self.obj().initialize();
        }
    }

    impl WidgetImpl for ImageDropPagePriv {}
    impl BinImpl for ImageDropPagePriv {}

    impl ImageDropPagePriv {}
}

glib::wrapper! {
    pub struct ImageDropPage(ObjectSubclass<imp::ImageDropPagePriv>)
    @extends gtk::Widget, adw::Bin;
}

impl ImageDropPage {
    pub fn new() -> ImageDropPage {
        let color_panel: ImageDropPage = glib::Object::builder::<ImageDropPage>().build();
        color_panel
    }

    fn initialize(&self) {
        self.imp().file_verified.set(false);
        self.setup_drop_target();
    }

    fn setup_drop_target(&self) {
        let imp = self.imp();

        let formats = gdk::ContentFormats::new(&["text/uri-list"]);
        let drop_target = gtk::DropTargetAsync::new(Some(formats), gdk::DragAction::COPY);

        drop_target.connect_accept(
            clone!(@strong self as this => move |_drop_target, drop_value| {
                this.imp().file_verified.set(false);
                let formats = drop_value.formats();
                if contain_mime_type(formats) {
                    drop_value.read_value_async(gio::File::static_type(), glib::PRIORITY_DEFAULT, None::<&gio::Cancellable>, 
                    clone!(@strong this => move |value| {
                            this.verify_file_valid(value)
                        })
                    );
                    return true
                }
                return false
            })
        );

        drop_target.connect_drop(
            clone!(@strong self as this => move |_drop_target, drop_value, _x, _y| {
                drop_value.read_value_async(gio::File::static_type(), glib::PRIORITY_DEFAULT, None::<&gio::Cancellable>, 
                    clone!(@strong this => move |value| {
                        if !this.imp().file_verified.get() {
                            add_error_toast(i18n("Unable to verify file on drop, try with the file chooser in the upper left corner."))
                        }
                        this.load_value_async(value)
                    })
                );
                return true
            }),
        );

        imp.overlay.add_controller(drop_target);
    }

    fn verify_file_valid(&self, value: Result<glib::Value, glib::Error>) {
        match value {
            Ok(value) => {
                let file: gio::File = value.get::<gio::File>().unwrap();
                let uri = file.path().unwrap();
                self.imp().file_verified.set(Path::new(&uri).exists());
            },
            Err(e) => {
                error!("{}", e);
                return;
            }
        }

    }

    fn load_value_async(&self, value: Result<glib::Value, glib::Error>) {
        match value {
            Ok(value) => {
                let file: gio::File = value.get::<gio::File>().unwrap();
                let uri = file.path().unwrap();
                self.load_image(uri.to_str().unwrap());
                debug!("{:?}", uri);
            },
            Err(e) => {
                add_error_toast(i18n("Unable to read drop."));
                error!("{}", e);
                return;
            }
        }
    }

    pub fn load_image(&self, uri: &str) -> bool {
        let dropped_image = DroppedImage::new();
        if let Err(e) = dropped_image.load_image(uri) {
            error_image_toast(uri);
            error!("{}", e);
            return false
        } else {
            let imp = self.imp();
            imp.thief_panel.set_image(dropped_image);
            imp.status.hide();
            open_image_toast(uri);
            return true;
        }
    }


    fn set_image(&self, image: DroppedImage) {
        let imp = self.imp();
        self.imp().thief_panel.set_visible(true);
        imp.image_bin.set_child(Some(&image));
        self.imp().thief_panel.imp().image.replace(Some(image));
        self.imp().thief_panel.list_store().remove_all();
        self.imp().thief_panel.start_extraction();
    }

}

fn contain_mime_type(formats: ContentFormats) -> bool {
    let mimes = ["text/uri-list"];
    mimes.iter().any(|m| formats.contain_mime_type(m))
}